dask/array/gufunc.py

```diff
@@ -0,0 +1,343 @@
+        if len(output_dtypes) != nout:
+            raise ValueError("Must specify tuple of %d dtypes for `output_dtypes` for function with multiple outputs" % nout)
+    if nout is None and (isinstance(output_dtypes, tuple) or isinstance(output_dtypes, list)):
+        raise ValueError("Must specify single dtype for `output_dtypes` for function with one output")
+
+
+    ## Miscellaneous
+    if output_sizes is None:
+        output_sizes = {}
+
+    # Main code:
+    ## Cast all input arrays to dask
+    args = [asarray(a) for a in args]
+
+    if len(core_input_dimss) != len(args):
+        ValueError("According to `signature`, `func` requires %d arguments, but %s given"
+                   % (len(core_output_dimss), len(args)))
+
+    ## Assess input args for loop dims
+    input_shapes = [a.shape for a in args]
+    num_loopdims = [len(s) - len(cd) for s, cd in zip(input_shapes, core_input_dimss)]
+    max_loopdims = max(num_loopdims) if num_loopdims else None
+    _core_input_shapes = [dict(zip(cid, s[n:])) for s, n, cid in zip(input_shapes, num_loopdims, core_input_dimss)]
+    core_shapes = merge(output_sizes, *_core_input_shapes)
+
+    loop_input_dimss = [tuple("__loopdim%d__" % d for d in range(max_loopdims - n, max_loopdims)) for n in num_loopdims]
+    loop_input_shapes = [s[:n] for s, n in zip(input_shapes, num_loopdims)]
+
+    input_dimss = [l + c for l, c in zip(loop_input_dimss, core_input_dimss)]
+
+    loop_output_dims = max(loop_input_dimss, key=len) if loop_input_dimss else set()
+    loop_output_shape = broadcast_shapes(*loop_input_shapes)
+
+    ## Prepare function
+    if vectorize:
+        signature = compile_signature(core_input_dimss, core_output_dimss)
+        ufunc = np_vectorize(func, signature=signature)
+    else:
+        # Wrapper that broadcasts only loop dims (as opposed to numpy.broadcast_arrays)
+        def ufunc(*args, **kwargs):
+            input_shapes = (a.shape for a in args)
+            loop_input_blockshapes = (s[:n] for s, n in zip(input_shapes, num_loopdims))
+            loop_output_blockshape = broadcast_shapes(*loop_input_blockshapes)
+            shapes = (loop_output_blockshape + a.shape[n:] for a, n in zip(args, num_loopdims))
+            bargs = (np_broadcast_to(a, s) for a, s in zip(args, shapes))
+            return func(*bargs, **kwargs)
+
+    ## Apply function - use atop here
+    arginds = list(concat(zip(args, input_dimss)))
+
+    ### Treat direct output
+    if nout is None:
+        core_output_dimss = [core_output_dimss]
+        output_dtypes = [output_dtypes]
+
+    ### Use existing `atop` but only with loopdims to enforce
+    ### concatenation, for coredims, that appear also at the output
+    ### Modifying `atop` could improve things here.
+    tmp = atop(ufunc, loop_output_dims, *arginds,
+               dtype=int,  # Only dummy dtype, anyone will do
+               concatenate=concatenate,
+               **kwargs)
```

Comment:
Given that we don't really want an array output, would it make sense to drop down to `top` here?

I'm also open to teaching `atop` how to handle multiple outputs.  (I'll write a longer note about this in the main comments)

Reply:
I also needed `unify_chunks` being called, so that I basically use 80% of `atop` code here. After below's discussion potential refactor can be continued.